### ISSUE TITLE
trouble building with docker for Dockerfile.arm64v8

### ISSUE THREAD

**Issue by rhildred** [coderaiser/cloudcmd#291]
* **Version** (`cloudcmd -v`): checkout from github 2020-06-06 (4076ded)
* **Node Version** `node -v`: v10.19.0
* **OS** (`uname -a` on Linux): Linux rivendale.selab.ca 5.4.0-1011-raspi #11-Ubuntu SMP Fri May 8 07:43:33 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux
* **Browser name/version**: n/a
* **Used Command Line Parameters**: n/a
* **Changed Config**: when I first tried with arm64v8/node:slim the npm install --production core dumped and exited 134
I am building a computer lab appliance and want the student containers to have Ubuntu so I tried again with Ubuntu:

```
< FROM arm64v8/node:slim
---
> FROM arm64v8/ubuntu
10c10
< RUN apt-get update && apt-get -y install python3 make g++
---
> RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get -y install nodejs npm python3 make g++
14,17c14

``` 
This is the error that I get:

```
Traceback (most recent call last):
  File "/usr/share/nodejs/node-gyp/gyp/gyp_main.py", line 6, in <module>
    from pkg_resources import load_entry_point
ModuleNotFoundError: No module named 'pkg_resources'
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/share/nodejs/node-gyp/lib/configure.js:355:16)
gyp ERR! stack     at ChildProcess.emit (events.js:198:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:248:12)
gyp ERR! System Linux 5.4.0-1011-raspi
gyp ERR! command "/usr/bin/node" "/usr/bin/node-gyp" "rebuild"
gyp ERR! cwd /usr/src/app/node_modules/node-pty
gyp ERR! node -v v10.19.0
gyp ERR! node-gyp -v v6.1.0
gyp ERR! not ok 
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! node-pty@0.9.0 install: `node scripts/install.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the node-pty@0.9.0 install script.
```

I also changed the docker-compose.yml to build the Dockerfile

```
   web:
+    build: .
     ports:
       - 8000:8000

```

I am really eager for this to work on PI. I have 2 virtual classes scheduled for the fall term in python and Jekyl. It looks like I maybe should change the version of node-pty. I will keep digging in to this myself but I was hoping that you might have some insight.

Thanks

Rich

**Comment by coderaiser** [comment]
Rich thank you for an issue :), @timoknapp, could you please help with this :)?

**Comment by timoknapp** [comment]
Hi @rhildred ,

What are actually trying to do? Are you depending on the official `cloudcmd` docker images or are you trying to build it yourself? It looks to me, that second is the case. If you are building your own `Docker image` for `cloudcmd`. Some things have to be kept in mind:

1. You need to depend on a proper `node` based image, which is supported by your architecture of choice. In case of a Raspberry you are fine with the `arm64v8/node:slim`.
2. A `python`installation is necessary for the build, cause `node-pty` is required in the build process.
3. It is important that the `image` is build on the relevant system architecture. If you want to eventually run it on a Raspberry Pi (`arm32v7` or `arm64v8`), it has to be built on such an architecture too.

In general there are currently issues with the official releases of `cloudcmd` for `arm`-based Docker images on `Travis CI` because of the 3rd bullet point.
You can try if [this](https://hub.docker.com/r/team0/rpi-cloudcmd) image is working for you. It has been built on a  Raspberry Pi 4

If you need further assistance, let me know :)

**Comment by rhildred** [comment]
Hi @timoknapp and @coderaiser 

Thanks so much for your quick response. I did try the docker-compose yml with no build:.. It said it couldn't find an arm64/V8 image. 
My end design is to use cloudcmd as a middleware in my own container that is a complete development environment for a programming class in Jekyll, Python or WordPress. I had succesful pilots with this using Theia. Theia is too big and slow for a pi I feel. I also can't get my Theia containers from my pilots to build any more.

I did build the container on the pi. I will try your image and some more investigation on node-pty. I will also post my fix when I get one. 

Thanks so much for your help

Rich

Thanks

I am buil

**Comment by timoknapp** [comment]
@rhildred 

Okay sounds good. Feedback is always appreciated. Hope I'll manage to publish the fix for the `docker manifests` soon though.

Greetings
Timo

**Comment by coderaiser** [comment]
@timoknapp, @rhildred any progress on this :)?

**Comment by timoknapp** [comment]
Hi @coderaiser, 

I havent had a chance to properly update the `docker manifest` yet. But its still on my bucket-list. ✌️ 

@rhildred: Have you had a chance to resolve you initial issue? If so, can you close the issue?

**Comment by timoknapp** [comment]
Hi @coderaiser ,

The time has come, I plan to propose the PR with regards to the `docker manifest` this month. In fact the ongoing [Hacktoberfest](https://hacktoberfest.digitalocean.com) just reminded me about it.

Cheers,
Timo

**Comment by coderaiser** [comment]
@timoknapp that's amazing :).

**Comment by coderaiser** [comment]
@timoknapp any progress on this?

**Comment by timoknapp** [comment]
Hi @coderaiser ,

could not finish yet. But there were quite some changes within `docker` on the multi-architecture builds:

https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/

I guess the simplest solution would be using the new GitHub Action for the docker builds instead of using `travis`. It would look similar to the parts under "Let’s go to production" from the link above. Furthermore there is an [official GitHub Action](https://github.com/docker/setup-buildx-action) from `docker` by now: 
Technically it could also work with `travis` similar to [this](https://github.com/jdrouet/docker-on-ci/blob/master/.travis.yml), even though it looks more custom made.

Whats your preferred solution?

**Comment by coderaiser** [comment]
@timoknapp I think the simplest one is better :). I don't mind to switch to `actions` for docker builds, anyways there is already [node ci configured](https://github.com/coderaiser/cloudcmd/actions?query=workflow%3A%22Node+CI%22). On other hand there is a a lot code in `travis` configs, so if we switch would be amazing to remove all not used code.

Less code is better, so let's do it :).

**Comment by timoknapp** [comment]
Sounds good to me. I just updated my fork and added a simple action to try things out.
Lets hope the best 🙏

**Comment by timoknapp** [comment]
I tried it out and looks good so far. Just pushed a new multi-architecture build to docker hub, using my `docker` user. 

https://hub.docker.com/r/team0/cloudcmd-multi/tags?page=1&ordering=last_updated

In order make it fully running, you would need to provide the corresponding `docker` credentials as secrets within your repo settings. I added a `TODO:` for you inside the `docker.yml` workflow file. I will create a new PR whenever I am confident with the solution.
Cheers

**Comment by coderaiser** [comment]
@timoknapp that's sounds good :).

**Comment by timoknapp** [comment]
Hi @rhildred 

I guess this bug will be resolved whenever the new version of `cloudcmd`is released by @coderaiser .

Just as hint the following types of `docker` images are then [available](https://github.com/coderaiser/cloudcmd#docker):
| Architecture   | Type         |
|----------------|--------------|
| amd64          | linux        |
| arm/v7         | linux        |
| arm64 (arm/v8) | linux        |
| amd64          | linux-alpine |
| arm/v7         | linux-alpine |
| arm64 (arm/v8) | linux-alpine |

Regards
Timo

**Comment by coderaiser** [comment]
Landed in [v15.3.4](https://github.com/coderaiser/cloudcmd/releases/tag/v15.3.4) 🎉 , @timoknapp thanks a lot :)
@rhildred, is it works for you?
